File: android/app/src/main/java/local/gestorweb/android/BrowserActivity.java
package local.gestorweb.android;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.HttpAuthHandler;
import android.webkit.PermissionRequest;
import android.webkit.SslErrorHandler;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebStorage;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.webkit.ProxyConfig;
import androidx.webkit.ProxyController;
import androidx.webkit.Profile;
import androidx.webkit.WebViewCompat;
import androidx.webkit.WebViewFeature;
import java.io.ByteArrayInputStream;
import java.net.Authenticator;
import java.net.PasswordAuthentication;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;

public final class BrowserActivity extends AppCompatActivity {
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Executor direct = Runnable::run;
    private WebView webView;
    private EditText address;
    private ProgressBar progress;
    private ProfileStore.Profile profile;
    private LicenseStore licenses;
    private HistoryStore history;
    private boolean proxyApplied;
    private Profile webProfile;

    private final Runnable licenseCheck = new Runnable() {
        @Override public void run() {
            new Thread(() -> {
                LicenseClient.Result result = LicenseClient.verify(BrowserActivity.this, licenses.license(), licenses.deviceId());
                runOnUiThread(() -> {
                    if (!result.active) { licenses.clearLicense(); startActivity(new Intent(BrowserActivity.this, MainActivity.class).addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)); finish(); }
                    else handler.postDelayed(this, 60_000);
                });
            }).start();
        }
    };

    @SuppressLint("SetJavaScriptEnabled")
    @Override protected void onCreate(Bundle state) {
        super.onCreate(state);
        setContentView(R.layout.activity_browser);
        licenses = new LicenseStore(this); history = new HistoryStore(this);
        profile = new ProfileStore(this).findById(getIntent().getStringExtra("profileId"));
        if (profile == null || licenses.license().isEmpty()) { finish(); return; }
        address = findViewById(R.id.browserAddress); progress = findViewById(R.id.browserProgress); webView = findViewById(R.id.browserWebView);
        if (WebViewFeature.isFeatureSupported(WebViewFeature.MULTI_PROFILE)) {
            WebViewCompat.setProfile(webView, "gw_" + profile.id.replaceAll("[^a-zA-Z0-9_]", "_"));
            webProfile = WebViewCompat.getProfile(webView);
        }
        ((TextView)findViewById(R.id.browserProfileName)).setText(profile.name);
        ((TextView)findViewById(R.id.browserRoute)).setText(profile.torMode ? "TOR · ORBOT" : profile.hasProxy() ? profile.proxyType + " · PROXY" : "DIRECTO");
        findViewById(R.id.browserBack).setOnClickListener(v -> { if (webView.canGoBack()) webView.goBack(); });
        findViewById(R.id.browserForward).setOnClickListener(v -> { if (webView.canGoForward()) webView.goForward(); });
        findViewById(R.id.browserReload).setOnClickListener(v -> webView.reload());
        address.setOnEditorActionListener((v, action, event) -> { navigate(address.getText().toString()); return true; });
        configureWebView(); applyProxyAndOpen(); handler.postDelayed(licenseCheck, 60_000);
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void configureWebView() {
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true); settings.setDomStorageEnabled(!profile.inMemory); settings.setDatabaseEnabled(!profile.inMemory);
        settings.setGeolocationEnabled(false); settings.setSaveFormData(false); settings.setMediaPlaybackRequiresUserGesture(true);
        settings.setMixedContentMode(profile.forceHttps || !profile.compatMode ? WebSettings.MIXED_CONTENT_NEVER_ALLOW : WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE);
        settings.setCacheMode(profile.inMemory ? WebSettings.LOAD_NO_CACHE : WebSettings.LOAD_DEFAULT);
        settings.setUserAgentString(profile.effectiveUserAgent());
        CookieManager cookies = webProfile != null ? webProfile.getCookieManager() : CookieManager.getInstance(); cookies.setAcceptCookie(!profile.inMemory); cookies.setAcceptThirdPartyCookies(webView, profile.compatMode && !profile.antiLeak);

        String documentStart = SpoofInjector.buildScript(profile);
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            WebViewCompat.addDocumentStartJavaScript(webView, documentStart, Collections.singleton("*"));
        }

        webView.setWebChromeClient(new WebChromeClient() {
            @Override public void onProgressChanged(WebView view, int value) { progress.setProgress(value); progress.setVisibility(value >= 100 ? View.GONE : View.VISIBLE); }
            @Override public void onPermissionRequest(PermissionRequest request) { if (profile.antiLeak || profile.webrtcBlock) request.deny(); else request.grant(request.getResources()); }
        });
        webView.setWebViewClient(new WebViewClient() {
            @Override public void onPageStarted(WebView view, String url, Bitmap icon) {
                address.setText(url); progress.setVisibility(View.VISIBLE);
                history.add(profile, "GET", url, "navigation");
                if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) view.evaluateJavascript(documentStart, null);
            }
            @Override public void onPageFinished(WebView view, String url) { address.setText(url); progress.setVisibility(View.GONE); }
            @Override public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                String original = request.getUrl().toString(); String safe = normalizeUrl(original);
                if (!safe.equals(original)) { loadWithProfileHeaders(safe); return true; }
                return false;
            }
            @Override public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
                String url = request.getUrl().toString();
                if (profile.harEnabled) history.add(profile, request.getMethod(), url, request.isForMainFrame() ? "document" : "resource");
                if (profile.blockTrackers && TrackerBlocker.blocked(url)) return new WebResourceResponse("text/plain", "utf-8", new ByteArrayInputStream(new byte[0]));
                return null;
            }
            @Override public void onReceivedHttpAuthRequest(WebView view, HttpAuthHandler auth, String host, String realm) {
                if (profile.hasProxyAuth()) auth.proceed(profile.proxyUser, profile.proxyPass); else super.onReceivedHttpAuthRequest(view, auth, host, realm);
            }
            @Override public void onReceivedSslError(WebView view, SslErrorHandler ssl, SslError error) {
                if (profile.hasProxy() || profile.torMode) ssl.proceed(); else ssl.cancel();
            }
        });
    }

    private void applyProxyAndOpen() {
        String route = profile.torMode ? "socks://127.0.0.1:9050" : profile.proxyUri();
        if (route.isEmpty()) { navigate(profile.url); return; }
        if (profile.hasProxyAuth()) Authenticator.setDefault(new Authenticator() {
            @Override protected PasswordAuthentication getPasswordAuthentication() { return new PasswordAuthentication(profile.proxyUser, profile.proxyPass.toCharArray()); }
        });
        if (!WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE)) { Toast.makeText(this, "Este WebView no soporta proxy override", Toast.LENGTH_LONG).show(); navigate(profile.url); return; }
        try {
            ProxyConfig config = new ProxyConfig.Builder().addProxyRule(route).build();
            ProxyController.getInstance().setProxyOverride(config, direct, () -> { proxyApplied = true; navigate(profile.url); });
        } catch (Exception error) { Toast.makeText(this, "Proxy no compatible: " + error.getMessage(), Toast.LENGTH_LONG).show(); navigate("about:blank"); }
    }

    private void navigate(String raw) { loadWithProfileHeaders(normalizeUrl(raw == null || raw.trim().isEmpty() ? "about:blank" : raw.trim())); }
    private String normalizeUrl(String raw) {
        String value = raw;
        if (!value.equals("about:blank") && Uri.parse(value).getScheme() == null) value = "https://" + value;
        if (profile.stripTrackingParams) value = TrackerBlocker.stripTracking(value);
        if (profile.forceHttps && value.startsWith("http://")) value = "https://" + value.substring(7);
        return value;
    }
    private void loadWithProfileHeaders(String url) {
        address.setText(url); Map<String,String> headers = new HashMap<>(); headers.put("DNT", "1"); headers.put("Sec-GPC", "1");
        if (profile.strictReferer) headers.put("Referer", "");
        webView.loadUrl(url, headers);
    }

    @Override public void finish() { super.finish(); overridePendingTransition(android.R.anim.fade_in, R.anim.slide_out_left); }
    @Override protected void onDestroy() {
        handler.removeCallbacks(licenseCheck);
        if (webView != null) {
            webView.stopLoading();
            if (profile != null && (profile.inMemory || profile.autoWipeClose)) {
                webView.clearHistory(); webView.clearCache(true); webView.clearFormData();
                CookieManager cookies = webProfile != null ? webProfile.getCookieManager() : CookieManager.getInstance(); cookies.removeAllCookies(null); cookies.flush();
                if (webProfile != null) webProfile.getWebStorage().deleteAllData(); else WebStorage.getInstance().deleteAllData();
            }
            webView.destroy();
        }
        if (proxyApplied && WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE)) ProxyController.getInstance().clearProxyOverride(direct, () -> {});
        Authenticator.setDefault(null); super.onDestroy();
    }
}

File: android/app/src/main/java/local/gestorweb/android/BrowserActivity.kt
package local.gestorweb.android

import android.annotation.SuppressLint
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.net.http.SslError
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.webkit.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.webkit.ProxyConfig
import androidx.webkit.ProxyController
import androidx.webkit.WebViewFeature
import com.google.android.material.bottomsheet.BottomSheetDialog

class BrowserActivity : AppCompatActivity() {

    private val handler = Handler(Looper.getMainLooper())
    private val direct  = java.util.concurrent.Executor { it.run() }

    private lateinit var webView:        WebView
    private lateinit var etAddress:      EditText
    private lateinit var progressBar:    ProgressBar
    private lateinit var btnBack:        Button
    private lateinit var btnForward:     Button
    private lateinit var tvSecurityIcon: TextView

    private lateinit var licenses: LicenseStore
    private var profile: ProfileStore.Profile? = null

    private val licenseCheck = object : Runnable {
        override fun run() {
            Thread {
                val result = LicenseClient.verify(this@BrowserActivity, licenses.license(), licenses.deviceId())
                runOnUiThread {
                    if (!result.active) {
                        licenses.clearLicense()
                        startActivity(Intent(this@BrowserActivity, MainActivity::class.java)
                            .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP))
                        overridePendingTransition(R.anim.slide_in_left, R.anim.slide_out_right)
                        finish()
                    } else handler.postDelayed(this, 60_000)
                }
            }.start()
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        licenses = LicenseStore(this)
        val profileId = intent.getStringExtra("profileId")
        profile = ProfileStore(this).list().firstOrNull { it.id == profileId }
        if (profile == null || licenses.license().isEmpty()) { finish(); return }
        val p = profile!!

        setContentView(R.layout.activity_browser)

        webView        = findViewById(R.id.webView)
        etAddress      = findViewById(R.id.etAddress)
        progressBar    = findViewById(R.id.browserProgress)
        btnBack        = findViewById(R.id.btnBack)
        btnForward     = findViewById(R.id.btnForward)
        tvSecurityIcon = findViewById(R.id.tvSecurityIcon)

        // Title bar
        findViewById<TextView>(R.id.tvBrowserProfileName).text = p.name
        val modeStr = getString(R.string.label_mode_prefix) + p.privacyMode +
            if (p.proxy.isEmpty()) getString(R.string.label_direct_suffix)
            else getString(R.string.label_proxy_suffix)
        findViewById<TextView>(R.id.tvBrowserProfileMode).text = modeStr

        val pill = findViewById<TextView>(R.id.tvBrowserModePill)
        pill.text = p.privacyMode.uppercase()
        pill.setBackgroundResource(
            if (p.privacyMode == ProfileStore.MODE_STRICT) R.drawable.bg_pill_danger
            else R.drawable.bg_pill_accent
        )

        // Botones
        findViewById<Button>(R.id.btnCloseBrowser).setOnClickListener {
            finish()
            overridePendingTransition(R.anim.slide_in_left, R.anim.slide_out_right)
        }
        btnBack.setOnClickListener    { if (webView.canGoBack())    webView.goBack() }
        btnForward.setOnClickListener { if (webView.canGoForward()) webView.goForward() }
        findViewById<Button>(R.id.btnRefresh).setOnClickListener { webView.reload() }

        // ── menú "más" ──────────────────────────────────────────────────────
        findViewById<Button>(R.id.btnMore).setOnClickListener { showMoreMenu() }

        etAddress.setOnEditorActionListener { _, _, _ -> navigate(etAddress.text.toString()); true }

        webView.setBackgroundColor(Color.rgb(10, 14, 22))
        configureWebView(p)
        applyProxyAndOpen(p)
        handler.postDelayed(licenseCheck, 60_000)
    }

    // ── bottom sheet menu ─────────────────────────────────────────────────────

    private fun showMoreMenu() {
        val sheet = BottomSheetDialog(this)
        val view  = layoutInflater.inflate(R.layout.sheet_browser_menu, null)
        sheet.setContentView(view)

        // Compartir URL
        view.findViewById<LinearLayout>(R.id.menuItemShare).setOnClickListener {
            val url = etAddress.text.toString()
            startActivity(Intent.createChooser(
                Intent(Intent.ACTION_SEND).apply {
                    type = "text/plain"
                    putExtra(Intent.EXTRA_TEXT, url)
                }, getString(R.string.menu_share)
            ))
            sheet.dismiss()
        }

        // Copiar URL
        view.findViewById<LinearLayout>(R.id.menuItemCopy).setOnClickListener {
            val url = etAddress.text.toString()
            (getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager)
                .setPrimaryClip(ClipData.newPlainText("URL", url))
            Toast.makeText(this, getString(R.string.msg_url_copied), Toast.LENGTH_SHORT).show()
            sheet.dismiss()
        }

        // Limpiar caché y cookies del perfil
        view.findViewById<LinearLayout>(R.id.menuItemClearData).setOnClickListener {
            webView.clearCache(true)
            webView.clearHistory()
            CookieManager.getInstance().removeAllCookies(null)
            CookieManager.getInstance().flush()
            Toast.makeText(this, getString(R.string.msg_cache_cleared), Toast.LENGTH_SHORT).show()
            sheet.dismiss()
        }

        // Abrir en navegador externo
        view.findViewById<LinearLayout>(R.id.menuItemOpenExternal).setOnClickListener {
            val url = etAddress.text.toString()
            try {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
            } catch (e: Exception) {
                Toast.makeText(this, getString(R.string.msg_no_browser), Toast.LENGTH_SHORT).show()
            }
            sheet.dismiss()
        }

        sheet.show()
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun configureWebView(p: ProfileStore.Profile) {
        webView.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            databaseEnabled   = true
            setSupportMultipleWindows(true)
            setGeolocationEnabled(false)
            @Suppress("DEPRECATION") saveFormData = false
            mixedContentMode = if (p.privacyMode == ProfileStore.MODE_COMPAT)
                WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE else WebSettings.MIXED_CONTENT_NEVER_ALLOW
            cacheMode = if (p.privacyMode == ProfileStore.MODE_STRICT)
                WebSettings.LOAD_NO_CACHE else WebSettings.LOAD_DEFAULT
            if (p.userAgent.isNotEmpty()) userAgentString = p.userAgent
        }
        CookieManager.getInstance().apply {
            setAcceptCookie(p.privacyMode != ProfileStore.MODE_STRICT)
            setAcceptThirdPartyCookies(webView, p.privacyMode == ProfileStore.MODE_COMPAT)
        }
        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                if (newProgress < 100) {
                    progressBar.visibility = View.VISIBLE
                    progressBar.progress   = newProgress
                } else {
                    progressBar.visibility = View.GONE
                }
                btnBack.alpha    = if (webView.canGoBack())    1.0f else 0.4f
                btnForward.alpha = if (webView.canGoForward()) 1.0f else 0.4f
            }
        }
        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest) = false
            override fun onPageFinished(view: WebView, url: String) {
                etAddress.setText(url)
                tvSecurityIcon.text = if (url.startsWith("https://"))
                    getString(R.string.label_secure) else getString(R.string.label_insecure)
                btnBack.alpha    = if (view.canGoBack())    1.0f else 0.4f
                btnForward.alpha = if (view.canGoForward()) 1.0f else 0.4f
            }
            override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
                handler.cancel()
            }
        }
    }

    private fun applyProxyAndOpen(p: ProfileStore.Profile) {
        if (p.proxy.isEmpty() || !WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE)) {
            navigate(p.url); return
        }
        val proxy  = if ("://" in p.proxy) p.proxy else "http://${p.proxy}"
        val config = ProxyConfig.Builder().addProxyRule(proxy).addDirect().build()
        ProxyController.getInstance().setProxyOverride(config, direct) { navigate(p.url) }
    }

    private fun navigate(raw: String?) {
        var value = raw?.trim()?.ifEmpty { "about:blank" } ?: "about:blank"
        if (value != "about:blank" && Uri.parse(value).scheme == null) value = "https://$value"
        etAddress.setText(value)
        webView.loadUrl(value)
    }

    override fun onDestroy() {
        handler.removeCallbacks(licenseCheck)
        webView.stopLoading(); webView.destroy()
        if (WebViewFeature.isFeatureSupported(WebViewFeature.PROXY_OVERRIDE))
            ProxyController.getInstance().clearProxyOverride(direct) {}
        super.onDestroy()
    }
}
